compiler/ir/backend.jvm/src/org/jetbrains/kotlin/backend/jvm/lower/AddContinuationLowering.kt

```diff
@@ -36,201 +36,172 @@ import org.jetbrains.kotlin.ir.expressions.impl.*
+        body = IrExpressionBodyImpl(startOffset, endOffset, IrErrorExpressionImpl(startOffset, endOffset, returnType, message))
     }
 
-    private fun transformSuspendLambdasIntoContinuations(irFile: IrFile) {
-        val inlineReferences = mutableSetOf<IrCallableReference<*>>()
-        val suspendLambdas = mutableMapOf<IrFunctionReference, SuspendLambdaInfo>()
-        irFile.acceptChildren(object : IrInlineReferenceLocator(context) {
-            override fun visitInlineReference(argument: IrCallableReference<*>) {
-                inlineReferences.add(argument)
-            }
+    protected fun IrFunction.addCompletionValueParameter(): IrValueParameter =
+        addValueParameter(SUSPEND_FUNCTION_COMPLETION_PARAMETER_NAME, continuationType())
 
-            override fun visitFunctionReference(expression: IrFunctionReference, data: IrDeclaration?) {
-                expression.acceptChildren(this, data)
-                if (expression.isSuspend && expression.shouldBeTreatedAsSuspendLambda() && expression !in inlineReferences) {
-                    val isRestricted = expression.symbol.owner.extensionReceiverParameter?.type?.classOrNull?.owner?.annotations?.any {
-                        it.type.classOrNull?.signature == IdSignature.PublicSignature("kotlin.coroutines", "RestrictsSuspension", null, 0)
-                    } == true
-                    suspendLambdas[expression] = SuspendLambdaInfo(expression, isRestricted)
-                }
-            }
+    protected fun IrFunction.continuationType(): IrType =
+        context.ir.symbols.continuationClass.typeWith(returnType).makeNullable()
+}
 
-            private fun IrFunctionReference.shouldBeTreatedAsSuspendLambda() =
-                origin == IrStatementOrigin.LAMBDA ||
-                        origin == IrStatementOrigin.ADAPTED_FUNCTION_REFERENCE ||
-                        origin == IrStatementOrigin.SUSPEND_CONVERSION
-        }, null)
+private fun IrFunction.capturesCrossinline(): Boolean {
+    var result = false
+    accept(object : IrElementVisitorVoid {
+        override fun visitElement(element: IrElement) {
+            if (!result) element.acceptChildren(this, null)
+        }
 
-        for (lambda in suspendLambdas.values) {
-            (lambda.function.parent as IrDeclarationContainer).declarations.remove(lambda.function)
+        override fun visitFunction(declaration: IrFunction) {
+            functions.add(declaration)
+            super.visitFunction(declaration)
+            functions.remove(declaration)
         }
+
+        override fun visitGetValue(expression: IrGetValue) {
+            result = result || (expression.isReadOfCrossinline() && expression.symbol.owner.parent !in functions)
+        }
+
+        private val functions = mutableSetOf<IrFunction>()
+    }, null)
+    return result
+}
+
+private class SuspendLambdaLowering(context: JvmBackendContext) : SuspendLoweringBase(context) {
```

Comment:
IMHO moving it to a separate file would be more readable (and also maybe `SuspendLoweringBase` too)

Reply:
Done. `SuspendLoweringBase` is too short for that I think, it should look fine with `SuspendLambdaLowering`.